handler.go

```diff
@@ -60,6 +63,49 @@ func (o Options) ReadOptions(key string, obj interface{}) error {
 	return json.Unmarshal(msg, obj)
 }
 
+// Stream expects an array of json elements and allows to process them sequentially
+// this helps when one needs to parse a large json without having any memory leaks.
+// Returns ErrEmpty on empty key or when there are no more elements.
+// Returns ErrState when the stream has finished/encountered a Decode error.
+func (o Options) Stream(key string) (func(obj interface{}) error, error) {
```

Comment:
I think the `Stream` interface could be improved to return only an iterator function.

```go
 func (o Options) Stream(key string) (loadNext func(dest interface{}) error) {
```
`Stream` function must first complete preparation steps before it can return an iterator. It can do this and instead of instantly returning an error, provide a callback that will return it. The benefit is that for the user of this API, this function always returns an iterator and you have one less error case to check:
```go
func (o Options) Stream(key string) (loadNext func(dest interface{}) error) {  
        if _, ok := o[key]; !ok {                                              
                return func(interface{}) error {                               
                        errors.Wrap(err, "foo")                                
                }                                                              
        }                                                                      
                                                                               
        if _, err := dec.Token(); err != nil {                                 
                return func(interface{}) error {                               
                        errors.Wrap(err, "bar")                                
                }                                                              
        }                                                                      
                                                                               
        // ...                                                                 
                                                                               
        return func(dest interface{}) error {                                  
                // ...                                                         
        }                                                                      
}            
```

Reply:
Yeah, I can return the error within the function, that's fair.